client/driver/docker.go

```diff
@@ -489,6 +501,55 @@ func (d *DockerDriver) Start(ctx *ExecContext, task *structs.Task) (DriverHandle
 	return h, nil
 }
 
+func (d *DockerDriver) Cleanup(_ *ExecContext, res *CreatedResources) error {
+	retry := false
+	var merr multierror.Error
+	for key, resources := range res.Resources {
+		switch key {
+		case dockerImageResKey:
+			for _, value := range resources {
+				if err := d.cleanupImage(value); err != nil {
+					if structs.IsRecoverable(err) {
+						// This will be retried so put it back into the map
+						res.Add(key, value)
```

Comment:
Ah, missed that. Fixed and tested.

The logic is now:
* Clear the entire map key first since errors should be rare, avoid constantly shuffling the slice
* On *any* error, add that key back to the map
* On recoverable errors, mark the returned error as recoverable

This means if you have 2 errors - 1 recoverable and 1 not - both will be retried. If both errors were unrecoverable, neither would be retried. I chose this logic so TaskRunner can choose how to report leaked resources whether they were due to a fatal or recoverable error.